### PR TITLE
[FLINK-10984] [Build System, BuildSystem / Shaded] add flink-shaded-h…

### PR DESCRIPTION
**Pull request by sunjincheng121**
add the `flink-shaded-hadoop` module to the `flink-shaded`.

### CODE REVIEW

File: flink-shaded-hadoop-2/pom.xml [hunk 1]
```diff
@@ -0,0 +1,977 @@
...
+	<properties>
+		<hadoop.version>2.4.1</hadoop.version>
+	</properties>
+
+	<dependencies>
+
+		<dependency>
+			<groupId>com.google.code.findbugs</groupId>
+			<artifactId>jsr305</artifactId>
+			<scope>provided</scope>
```
Comment (zentol) [inline flink-shaded-hadoop-2/pom.xml]: let's add a comment that this is provided by Flink

File: flink-shaded-hadoop-2/pom.xml [hunk 2]
```diff
@@ -0,0 +1,977 @@
...
+			<version>1.3.9</version>
+		</dependency>
+		<dependency>
+			<groupId>org.slf4j</groupId>
+			<artifactId>slf4j-api</artifactId>
+			<scope>provided</scope>
```
Comment (zentol) [inline flink-shaded-hadoop-2/pom.xml]: same as above

File: tools/releasing/deploy_staging_jars.sh [hunk 3]
```diff
@@ -35,9 +35,20 @@ if [[ `basename $CURR_DIR` != "tools" ]] ; then
...
-$MVN clean deploy -Prelease -DskipTests -DretryFailedDeploymentCount=10
+$MVN clean deploy $COMMON_OPTIONS $EXCLUDE_HADOOP_MODULE
+
+HADOOP_MODULE="-pl flink-shaded-hadoop-2"
+
+HADOOP_VERSIONS=("2.4.1" "2.6.5" "2.7.5" "2.8.3")
 
+for i in "${!HADOOP_VERSIONS[@]}"; do
+echo "Deploying flink-shaded-hadoop $HADOOP_VERSIONS[$i] version"
+    $MVN clean deploy gs$COMMON_OPTIONS $HADOOP_MODULE "-Dhadoop.version=${HADOOP_VERSIONS[$i]}"
```
Comment (zentol) [inline tools/releasing/deploy_staging_jars.sh]: remove `gs` after `deploy`

File: tools/releasing/deploy_staging_jars.sh [hunk 4]
```diff
@@ -35,9 +35,20 @@ if [[ `basename $CURR_DIR` != "tools" ]] ; then
...
 echo "Deploying to repository.apache.org"
-$MVN clean deploy -Prelease -DskipTests -DretryFailedDeploymentCount=10
+$MVN clean deploy $COMMON_OPTIONS $EXCLUDE_HADOOP_MODULE
+
+HADOOP_MODULE="-pl flink-shaded-hadoop-2"
+
+HADOOP_VERSIONS=("2.4.1" "2.6.5" "2.7.5" "2.8.3")
 
+for i in "${!HADOOP_VERSIONS[@]}"; do
+echo "Deploying flink-shaded-hadoop $HADOOP_VERSIONS[$i] version"
```
Comment (zentol) [inline tools/releasing/deploy_staging_jars.sh]: let's remove this line; the mvn output will make it obvious what is happening

File: flink-shaded-hadoop-2/pom.xml [hunk 5]
```diff
@@ -0,0 +1,977 @@
...
+						<phase>none</phase>
+					</execution>
+					<execution>
+						<id>shade-hadoop</id>
+						<phase>package</phase>
+						<goals>
+							<goal>shade</goal>
+						</goals>
+						<configuration>
+							<shadedArtifactAttached>false</shadedArtifactAttached>
```
Comment (zentol) [inline flink-shaded-hadoop-2/pom.xml]: This is the default, so we can omit it like in other modules.

File: flink-shaded-hadoop-2/src/main/resources/META-INF/NOTICE [hunk 6]
```diff
@@ -0,0 +1,31 @@
+flink-shaded-hadoop2
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-shaded-hadoop-2/src/main/resources/META-INF/NOTICE]: I just follow the `flink-shaded-asm6` style. do you think we need to add a dash? If so, do we need to add a dash for `flink-shaded-asm61`?

File: flink-shaded-hadoop-2/pom.xml (hunk 5 above)
Comment (sunjincheng121) [reply flink-shaded-hadoop-2/pom.xml]: Yes, that's true.

File: flink-shaded-hadoop-2/pom.xml [hunk 7]
```diff
@@ -0,0 +1,977 @@
...
+								<relocation>
+									<pattern>org.htrace</pattern>
+									<shadedPattern>org.apache.flink.hadoop.shaded.org.htrace</shadedPattern>
+								</relocation>
+								<relocation>
+									<pattern>org.codehaus.jackson</pattern>
+									<shadedPattern>org.apache.flink.hadoop.shaded.org.codehaus.jackson</shadedPattern>
+								</relocation>
+							</relocations>
+							<transformers>
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-shaded-hadoop-2/pom.xml]: Yes, you are right!

File: flink-shaded-hadoop-2/pom.xml [hunk 8]
```diff
@@ -0,0 +1,977 @@
...
+					</execution>
+				</executions>
+			</plugin>
+			<!-- Relocate the Hadoop's Guava dependency into a different namespace and put Hadoo into a fat-jar.-->
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-shade-plugin</artifactId>
+				<executions>
+					<execution>
+						<!--Disable shade-flink execution to avoid duplicate relocation of dependencies-->
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-shaded-hadoop-2/pom.xml]: Ohh, we can remove this execution. right?

**Comment by sunjincheng121** [comment]
Thanks for your review @zentol!
I had uploaded the PR accordingly.   And only one question left about whether we should add a dash for `flink-shaded-hadoop2`.  :)
Best,
Jincheng

File: flink-shaded-hadoop-2/src/main/resources/META-INF/NOTICE (hunk 6 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-shaded-hadoop-2/src/main/resources/META-INF/NOTICE]: yes, asm6 should also have the dash. let's add this as a hotfix.

File: flink-shaded-hadoop-2/pom.xml [hunk 9]
```diff
@@ -0,0 +1,968 @@
...
+					</execution>
+				</executions>
+			</plugin>
+			<!-- Relocate the Hadoop's Guava dependency into a different namespace and put Hadoo into a fat-jar.-->
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-shade-plugin</artifactId>
+				<executions>
+					<execution>
+						<id>shade-hadoop</id>
```
Comment (zentol) [inline flink-shaded-hadoop-2/pom.xml]: should be called `shade-flink` so that we inherit from the root pom.

**Comment by sunjincheng121** [comment]
Great thanks for your quick reply @zentol 
I have had updated the PR!  I appreciate if you can have the other look! :)

**Pull request merged**